cirq/protocols/apply_unitary_to_tensor.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Union, Sequence, TypeVar
+
+import numpy as np
+from typing_extensions import Protocol
+
+from cirq import linalg
+from cirq.protocols.unitary import unitary
+
+
+# This is a special indicator value used by the apply_unitary_to_tensor method
+# to determine whether or not the caller provided a 'default' argument. It must
+# be of type np.ndarray to ensure the method has the correct type signature in
+# that case. It is checked for using `is`, so it won't have a false positive if
+# the user provides a different np.array([]) value.
+RaiseTypeErrorIfNotProvided = np.array([])  # type: np.ndarray
+
+TDefault = TypeVar('TDefault')
+
+
+class SupportsApplyUnitaryToTensor(Protocol):
```

Comment:
This use of Tensor for a tensor of a special shape bothers me.  TensorState instead of Tensor maybe?  (elsewhere we call the wavefunction the state)

Reply:
The tensor is sometimes a unitary matrix. It may not be wavefunction.